main.go

```diff
@@ -0,0 +1,176 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"path"
+
+	"github.com/web-platform-tests/wpt.fyi/shared"
+	"gopkg.in/yaml.v3"
+)
+
+func main() {
+	resolveMergeConflict()
+}
+
+func resolveMergeConflict() {
+	pendingMetadata := getPendingMetadata()
+	for testName, links := range pendingMetadata {
+		for _, link := range links {
+			toWPTMetadata(testName[1:], link)
+		}
+	}
+}
+
+// toWPTMetadata converts a pending metadata entry to a wpt-metadata META.yml
+// entry and writes it to disk.
+//
+// Directories and META.yml files are created as needed, or an existing one may
+// be updated.
+func toWPTMetadata(testName string, link shared.MetadataLink) {
+	dirpath := path.Dir(testName)
+	testpath := path.Base(testName)
+	createDirIfNeeded(dirpath)
+
+	// Open the META.yml file that we would write to, if it exists. In the case
+	// where it does exist we don't want to overwrite the file, so open in
+	// read-write-or-create mode.
+	filepath := dirpath + "/META.yml"
+	f, err := os.OpenFile(filepath, os.O_RDWR|os.O_CREATE, 0755)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f.Close()
+
+	fi, err := f.Stat()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if fi.Size() == 0 {
+		// This is an entirely new META.yml file, so create a new
+		// shared.Metadata representation for our expectation and write it out.
+		fmt.Println("CREATED-META-YML: for test", testName, "with url", link.URL)
+		metadata := shared.Metadata{
+			Links: []shared.MetadataLink{
+				{
+					Product: link.Product,
+					URL:     link.URL,
+					Results: []shared.MetadataTestResult{{
+						TestPath: testpath,
+					}},
+				},
+			},
+		}
+		writeToFile(metadata, f)
+		return
+	}
+
+	data, err := ioutil.ReadAll(f)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var metadata shared.Metadata
+	err = yaml.Unmarshal(data, &metadata)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// First check whether we have an existing crbug link in this META.yml; if
+	// so we just want to append our testname to the set of tests associated
+	// with it.
+	hasAdded := false
+	for index, existingLink := range metadata.Links {
+		if !link.Product.MatchesProductSpec(existingLink.Product) {
+			continue
+		}
+
+		// Check whether we are already listed in this entry.
+		for _, result := range existingLink.Results {
+			if result.TestPath == testpath {
+				fmt.Println("EXISTING-LINK: for test", testName, "with url", link.URL)
+				return
```

Comment:
Returning in the middle of this for loop, doesn't that mean that another link after this one could be new and is now going to be silently dropped on the floor? If this is a bug, can you add a test for this case?

To make it more obvious what's going on, would it be possible to structure this as looping through the smallest "atoms" of metadata we could add, and for each one check if the equivalent is already there and otherwise add it? Or would that require looping through all existing metadata for each addition because of how the data happens to be structured?

Reply:
Yea I simplified the code. The smallest `atoms` of metadata is [looped here](https://github.com/web-platform-tests/wpt-metadata/pull/1139/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R24). I will refactor and add tests